### PR TITLE
`azurerm_redhat_openshift_cluster` - support for the `preconfigured_network_security_group_enabled` property

### LINKED ISSUE #25059: azurerm_redhat_openshift_cluster : Support for preconfiguredNSG
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Description

Please add support for networkprofile.preconfiguredNSG  azurerm_redhat_openshift_cluster as shown in the bicep counterpart: https://learn.microsoft.com/en-us/rest/api/openshift/open-shift-clusters/update?view=rest-openshift-2023-09-04&tabs=HTTP#networkprofile

### New or Affected Resource(s)/Data Source(s)

azurerm_redhat_openshift_cluster

### Potential Terraform Configuration

```hcl
network_profile {
...
   preconfigured_nsg = true|false
}
```

### References

https://learn.microsoft.com/en-us/rest/api/openshift/open-shift-clusters/update?view=rest-openshift-2023-09-04&tabs=HTTP#networkprofile

### PR DESCRIPTION
**Pull request by scottd018**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave "+1" or "me too" comments, they generate extra noise for PR followers and do not help prioritize for review

## Description

This adds support for using a preconfigured network security group for Azure Red Hat OpenShift clusters.  This brings parity with the Azure CLI and the recently released BYO-NSG feature in ARO and allows users to consume this feature.  See https://learn.microsoft.com/en-us/azure/openshift/howto-bring-nsg for more details.

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

### Prior to Change

Prior to this change, the automation would exit with an error similar to the following, indicating that the `enabled-preconfigured-nsg` flag was not passed to tell the ARO resource provider that the BYO-NSG workflow needs to be enabled:

```
        API Response:
        
        ----[start]----
        {
            "id": "/subscriptions/xxxx/providers/microsoft.redhatopenshift/locations/eastus/operationsstatus/7738694c-c0f3-4008-87a8-316ea61d78d5",
            "name": "7738694c-c0f3-4008-87a8-316ea61d78d5",
            "status": "Failed",
            "startTime": "2024-06-03T13:58:11.433094028Z",
            "endTime": "2024-06-03T13:58:12.839247755Z",
            "error": {
                "code": "InvalidLinkedVNet",
                "message": "The provided subnet '/subscriptions/xxxx/resourceGroups/test-rg/providers/Microsoft.Network/virtualNetworks/test-vnet/subnets/main-subnet' is invalid: must not have a network security group attached.",
                "target": "properties.masterProfile.subnetId"
            }
        }
        
        -----[end]-----
```

### After Change

After the change, the test that was added in this PR named `TestAccOpenShiftCluster_preconfiguredNetworkSecurityGroup` successfully passes.  This test adds in the creation of a network security group, linking it to the subnets, passing in the `preconfigured_network_security_group_enabled` flag, and setting the appropriate permissions as required by the ARO RP:

```
make acctests SERVICE='redhatopenshift' TESTARGS='-run=TestAccOpenShiftCluster_preconfiguredNetworkSecurityGroup' TESTTIMEOUT='120m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/redhatopenshift -run=TestAccOpenShiftCluster_preconfiguredNetworkSecurityGroup -timeout 120m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccOpenShiftCluster_preconfiguredNetworkSecurityGroup
=== PAUSE TestAccOpenShiftCluster_preconfiguredNetworkSecurityGroup
=== CONT  TestAccOpenShiftCluster_preconfiguredNetworkSecurityGroup
--- PASS: TestAccOpenShiftCluster_preconfiguredNetworkSecurityGroup (3387.58s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/redhatopenshift       3389.881s
```

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `azurerm_redhat_openshift_cluster` - support for the `preconfigured_network_security_group_enabled property` [GH-{number}]

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes #25059

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

File: internal/services/redhatopenshift/redhat_openshift_cluster_resource.go [hunk 1]
```diff
@@ -193,6 +194,12 @@ func (r RedHatOpenShiftCluster) Arguments() map[string]*pluginsdk.Schema {
 							false,
 						),
 					},
+					"preconfigured_nsg_enabled": {
```
Comment (tombuildsstuff) [inline internal/services/redhatopenshift/redhat_openshift_cluster_resource.go:197]: To match other fields in the provider, can we spell this out as `network_security_group` here:

```suggestion
					"preconfigured_network_security_group_enabled": {
```

File: internal/services/redhatopenshift/redhat_openshift_cluster_resource.go [hunk 2]
```diff
@@ -57,9 +57,10 @@ type ClusterProfile struct {
 }
 
 type NetworkProfile struct {
-	OutboundType string `tfschema:"outbound_type"`
-	PodCidr      string `tfschema:"pod_cidr"`
-	ServiceCidr  string `tfschema:"service_cidr"`
+	OutboundType            string `tfschema:"outbound_type"`
+	PodCidr                 string `tfschema:"pod_cidr"`
+	ServiceCidr             string `tfschema:"service_cidr"`
+	PreconfiguredNSGEnabled bool   `tfschema:"preconfigured_nsg_enabled"`
```
Comment (tombuildsstuff) [inline internal/services/redhatopenshift/redhat_openshift_cluster_resource.go:63]: (as below)

```suggestion
	PreconfiguredNetworkSecurityGroupEnabled bool   `tfschema:"preconfigured_network_security_group_enabled"`
```

File: internal/services/redhatopenshift/redhat_openshift_cluster_resource_test.go [hunk 3]
```diff
@@ -565,6 +580,103 @@ resource "azurerm_redhat_openshift_cluster" "test" {
...
+
+  cluster_profile {
+    domain       = "aro-%[3]s.com"
+    version      = "4.13.23"
+  }
+
+  network_profile {
+    pod_cidr                  = "10.128.0.0/14"
+    service_cidr              = "172.30.0.0/16"
+    preconfigured_nsg_enabled = true
```
Comment (tombuildsstuff) [inline internal/services/redhatopenshift/redhat_openshift_cluster_resource_test.go:644]: (as above)

```suggestion
    preconfigured_network_security_group_enabled = true
```

File: website/docs/r/redhat_openshift_cluster.html.markdown [hunk 4]
```diff
@@ -217,6 +217,8 @@ A `network_profile` block supports the following:
 
 * `outbound_type` - (Optional) The outbound (egress) routing method. Possible values are `Loadbalancer` and `UserDefinedRouting`. Defaults to `Loadbalancer`. Changing this forces a new resource to be created.
 
+* `preconfigured_nsg_enabled` - (Optional) Whether a preconfigured network security group is being used on the subnets.  Defaults to `false`.  Changing this forces a new resource to be created.
```
Comment (tombuildsstuff) [inline website/docs/r/redhat_openshift_cluster.html.markdown:220]: (as above)

```suggestion
* `preconfigured_network_security_group_enabled` - (Optional) Whether a preconfigured network security group is being used on the subnets.  Defaults to `false`.  Changing this forces a new resource to be created.
```

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @scottd018 

Thanks for this PR - taking a look through here this mostly LGTM, if we can update `nsg` -> `network_security_group` to match other instances across the Provider and the tests pass then this should be good to go 👍 

Thanks!

**Comment by scottd018** [comment]
Sounds good. Thanks @tombuildsstuff ! That's exactly what I was after was feedback while BYO-NSG was broken so I can fix it, and then test it once the resource provider is fixed. Will make the requested changes in the meantime and I can post test results once I've had a chance to test.

Thanks again!

File: internal/services/redhatopenshift/redhat_openshift_cluster_resource.go (hunk 1 above)
Comment (scottd018) [reply internal/services/redhatopenshift/redhat_openshift_cluster_resource.go]: Apologies, I missed the fact that you made this change and I simply needed to commit.  Regardless, I fixed in commit becf28602323ab08a37def639367eb0975a74629

File: internal/services/redhatopenshift/redhat_openshift_cluster_resource.go (hunk 2 above)
Comment (scottd018) [reply internal/services/redhatopenshift/redhat_openshift_cluster_resource.go]: Apologies, I missed the fact that you made this change and I simply needed to commit.  Regardless, I fixed in commit becf28602323ab08a37def639367eb0975a74629

File: internal/services/redhatopenshift/redhat_openshift_cluster_resource_test.go (hunk 3 above)
Comment (scottd018) [reply internal/services/redhatopenshift/redhat_openshift_cluster_resource_test.go]: Apologies, I missed the fact that you made this change and I simply needed to commit.  Regardless, I fixed in commit becf28602323ab08a37def639367eb0975a74629

File: website/docs/r/redhat_openshift_cluster.html.markdown (hunk 4 above)
Comment (scottd018) [reply website/docs/r/redhat_openshift_cluster.html.markdown]: Apologies, I missed the fact that you made this change and I simply needed to commit.  Regardless, I fixed in commit becf28602323ab08a37def639367eb0975a74629

**Comment by scottd018** [comment]
@tombuildsstuff Resolved your suggestion and used the `terrafmt` tool to format a test that was improperly formatted, causing the pipeline run to fail.  Also updated the PR to support the new naming of the property.

Will wait for the upstream ARO RP to be fixed with NSG, run the tests, and inject the output into this PR.

**Review by tombuildsstuff — APPROVED** [review]
LGTM - thanks for making those changes @scottd018 

Since we're still waiting on the RP changes to roll out, I'm going to mark this as `blocked` for the moment, but once that's fixed let us know and we can run the tests/proceed with this one 👍

**Comment by scottd018** [comment]
@tombuildsstuff ARO RP is now fixed.  Tests have passed and the comment has been updated.  I had to push a new commit, as I forgot to add the 'Network Contributor' permissions to the NSG that I created in the test, so once that all passes we should be good.  Let me know if anything else is needed from me.  Thank you!

**Comment by scottd018** [comment]
@tombuildsstuff any updates on this?  Just checking.  Thank you!

**Comment by WadeBee-NG** [comment]
@tombuildsstuff Hi Tom, our team has been waiting for the fix. Any ETA on the PR getting merged?

**Review by stephybun — APPROVED** [review]
Tests look good, thanks @scottd018 LGTM 👍

**Pull request merged**